File: specs/002-hexagonal-arch-tdd/contracts/ports.ts
/**
 * Port Interfaces for Hexagonal Architecture
 *
 * These interfaces define the contracts between the domain/application layers
 * and the infrastructure adapters. They are owned by the core (domain/application)
 * and implemented by adapters.
 *
 * Generated for: 002-hexagonal-arch-tdd
 * Date: 2026-01-13
 */

// =============================================================================
// Domain Entities (referenced by ports)
// =============================================================================

export interface Document {
  id: string;
  filePath: string;
  title: string | null;
  pageCount: number | null;
  currentPage: number;
  scrollPosition: number;
  lastTtsChunkId: string | null;
  lastOpenedAt: string | null;
  fileHash: string | null;
  createdAt: string;
}

export interface Highlight {
  id: string;
  documentId: string;
  pageNumber: number;
  rects: Rect[];
  color: string;
  textContent: string | null;
  note: string | null;
  createdAt: string;
  updatedAt: string | null;
}

export interface Rect {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface VoiceInfo {
  id: string;
  name: string;
  language: string | null;
}

export interface TtsState {
  initialized: boolean;
  isSpeaking: boolean;
  isPaused: boolean;
  currentChunkId: string | null;
  currentVoice: VoiceInfo | null;
  rate: number;
}

export interface TtsInitResponse {
  available: boolean;
  backend: string | null;
  defaultVoice: string | null;
  error: string | null;
}

export interface TtsCapabilities {
  supportsUtterance: boolean;
  supportsRate: boolean;
  supportsPitch: boolean;
  supportsVolume: boolean;
}

// =============================================================================
// Input/Output DTOs
// =============================================================================

export interface HighlightCreate {
  documentId: string;
  pageNumber: number;
  rects: Rect[];
  color: string;
  textContent?: string;
}

export interface HighlightUpdate {
  color?: string;
  note?: string | null;
}

export interface BatchCreateResponse {
  highlights: Highlight[];
  created: number;
}

export interface DeleteResponse {
  success: boolean;
  deleted: number | null;
}

export interface ExportResponse {
  content: string;
  filename: string;
}

export interface FileExistsResponse {
  exists: boolean;
  filePath: string;
}

export type OrderBy = 'last_opened' | 'created' | 'title';
export type ExportFormat = 'markdown' | 'json' | 'text';

// =============================================================================
// Port Interfaces
// =============================================================================

/**
 * DocumentRepositoryPort
 *
 * Manages document persistence operations.
 * Implemented by: TauriDocumentRepository, MockDocumentRepository
 */
export interface DocumentRepositoryPort {
  /**
   * Add a new document to the library
   */
  add(filePath: string, title?: string, pageCount?: number): Promise<Document>;

  /**
   * Get a document by its content-hash ID
   */
  getById(id: string): Promise<Document | null>;

  /**
   * Get a document by its file path
   */
  getByPath(filePath: string): Promise<Document | null>;

  /**
   * List all documents with optional ordering and pagination
   */
  list(orderBy?: OrderBy, limit?: number, offset?: number): Promise<Document[]>;

  /**
   * Update reading progress for a document
   */
  updateProgress(
    id: string,
    page: number,
    scroll?: number,
    ttsChunkId?: string
  ): Promise<Document>;

  /**
   * Update document title
   */
  updateTitle(id: string, title: string): Promise<Document>;

  /**
   * Relocate a document to a new file path
   */
  relocate(id: string, newPath: string): Promise<Document>;

  /**
   * Remove a document from the library
   */
  remove(id: string): Promise<void>;

  /**
   * Check if a file exists at the given path
   */
  checkFileExists(filePath: string): Promise<FileExistsResponse>;
}

/**
 * HighlightRepositoryPort
 *
 * Manages highlight persistence operations.
 * Implemented by: TauriHighlightRepository, MockHighlightRepository
 */
export interface HighlightRepositoryPort {
  /**
   * Create a new highlight
   */
  create(input: HighlightCreate): Promise<Highlight>;

  /**
   * Create multiple highlights in a batch
   */
  batchCreate(inputs: HighlightCreate[]): Promise<BatchCreateResponse>;

  /**
   * Get a highlight by ID
   */
  getById(id: string): Promise<Highlight | null>;

  /**
   * List all highlights for a specific page
   */
  listForPage(documentId: string, pageNumber: number): Promise<Highlight[]>;

  /**
   * List all highlights for a document
   */
  listForDocument(documentId: string): Promise<Highlight[]>;

  /**
   * Update a highlight's color or note
   */
  update(id: string, input: HighlightUpdate): Promise<Highlight>;

  /**
   * Delete a highlight
   */
  delete(id: string): Promise<void>;

  /**
   * Delete all highlights for a document
   */
  deleteForDocument(documentId: string): Promise<DeleteResponse>;

  /**
   * Export highlights in various formats
   */
  export(documentId: string, format: ExportFormat): Promise<ExportResponse>;
}

/**
 * TtsPort
 *
 * Text-to-speech engine operations.
 * Implemented by: TauriNativeTtsAdapter, TauriAiTtsAdapter, MockTtsAdapter
 */
export interface TtsPort {
  /**
   * Initialize the TTS engine
   */
  init(): Promise<TtsInitResponse>;

  /**
   * List available voices
   */
  listVoices(): Promise<VoiceInfo[]>;

  /**
   * Speak a single text chunk
   */
  speak(text: string, chunkId?: string): Promise<void>;

  /**
   * Speak multiple chunks sequentially (for long content)
   */
  speakLong(chunks: Array<{ id: string; text: string }>): Promise<void>;

  /**
   * Stop all speech
   */
  stop(): Promise<void>;

  /**
   * Pause current speech
   */
  pause(): Promise<void>;

  /**
   * Resume paused speech
   */
  resume(): Promise<void>;

  /**
   * Set the active voice
   */
  setVoice(voiceId: string): Promise<void>;

  /**
   * Set playback rate (0.5 - 3.0)
   */
  setRate(rate: number): Promise<void>;

  /**
   * Get current TTS state
   */
  getState(): Promise<TtsState>;

  /**
   * Check engine capabilities
   */
  checkCapabilities(): Promise<TtsCapabilities>;
}

/**
 * SettingsPort
 *
 * User settings persistence.
 * Implemented by: TauriSettingsAdapter, MockSettingsAdapter
 */
export interface SettingsPort {
  /**
   * Get a setting value by key
   */
  get<T>(key: string): Promise<T | null>;

  /**
   * Set a setting value
   */
  set<T>(key: string, value: T): Promise<void>;

  /**
   * Get all settings
   */
  getAll(): Promise<Record<string, unknown>>;

  /**
   * Delete a setting
   */
  delete(key: string): Promise<void>;

  /**
   * Set multiple settings at once
   */
  setBatch(settings: Record<string, unknown>): Promise<void>;
}

/**
 * FileSystemPort
 *
 * File system operations (restricted by Tauri security model).
 * Implemented by: TauriFileSystemAdapter, MockFileSystemAdapter
 */
export interface FileSystemPort {
  /**
   * Read file contents
   */
  readFile(path: string): Promise<Uint8Array>;

  /**
   * Write file contents
   */
  writeFile(path: string, data: Uint8Array): Promise<void>;

  /**
   * Check if file exists
   */
  exists(path: string): Promise<boolean>;

  /**
   * Open file selection dialog
   */
  openDialog(options: DialogOptions): Promise<string | null>;

  /**
   * Open save file dialog
   */
  saveDialog(options: SaveDialogOptions): Promise<string | null>;
}

export interface DialogOptions {
  title?: string;
  filters?: Array<{ name: string; extensions: string[] }>;
  defaultPath?: string;
  multiple?: boolean;
}

export interface SaveDialogOptions {
  title?: string;
  filters?: Array<{ name: string; extensions: string[] }>;
  defaultPath?: string;
}

// =============================================================================
// Application Service Interfaces (Use Cases)
// =============================================================================

/**
 * LibraryService
 *
 * Use cases for document library management.
 */
export interface LibraryServicePort {
  openDocument(filePath: string): Promise<Document>;
  closeDocument(id: string): Promise<void>;
  getRecentDocuments(limit?: number): Promise<Document[]>;
  searchDocuments(query: string): Promise<Document[]>;
  removeDocument(id: string): Promise<void>;
}

/**
 * ReadingService
 *
 * Use cases for document reading and navigation.
 */
export interface ReadingServicePort {
  goToPage(documentId: string, page: number): Promise<void>;
  updateScroll(documentId: string, scroll: number): Promise<void>;
  bookmarkPosition(documentId: string): Promise<void>;
}

/**
 * TtsService
 *
 * Use cases for text-to-speech playback.
 */
export interface TtsServicePort {
  startReading(documentId: string, fromChunk?: string): Promise<void>;
  stopReading(): Promise<void>;
  togglePause(): Promise<void>;
  skipToChunk(chunkId: string): Promise<void>;
}

File: specs/002-hexagonal-arch-tdd/contracts/ports.rs
//! Port Traits for Hexagonal Architecture
//!
//! These traits define the contracts between the domain/application layers
//! and the infrastructure adapters. They are owned by the core (domain/application)
//! and implemented by adapters.
//!
//! Generated for: 002-hexagonal-arch-tdd
//! Date: 2026-01-13

use async_trait::async_trait;
use serde::{Deserialize, Serialize};

#[cfg(test)]
use mockall::automock;

// =============================================================================
// Domain Entities
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Document {
    pub id: String,
    pub file_path: String,
    pub title: Option<String>,
    pub page_count: Option<i32>,
    pub current_page: i32,
    pub scroll_position: f64,
    pub last_tts_chunk_id: Option<String>,
    pub last_opened_at: Option<String>,
    pub file_hash: Option<String>,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Highlight {
    pub id: String,
    pub document_id: String,
    pub page_number: i32,
    pub rects: Vec<Rect>,
    pub color: String,
    pub text_content: Option<String>,
    pub note: Option<String>,
    pub created_at: String,
    pub updated_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Rect {
    pub x: f64,
    pub y: f64,
    pub width: f64,
    pub height: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VoiceInfo {
    pub id: String,
    pub name: String,
    pub language: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TtsState {
    pub initialized: bool,
    pub is_speaking: bool,
    pub is_paused: bool,
    pub current_chunk_id: Option<String>,
    pub current_voice: Option<VoiceInfo>,
    pub rate: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TtsInitResponse {
    pub available: bool,
    pub backend: Option<String>,
    pub default_voice: Option<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TtsCapabilities {
    pub supports_utterance: bool,
    pub supports_rate: bool,
    pub supports_pitch: bool,
    pub supports_volume: bool,
}

// =============================================================================
// Input/Output DTOs
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HighlightCreate {
    pub document_id: String,
    pub page_number: i32,
    pub rects: Vec<Rect>,
    pub color: String,
    pub text_content: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HighlightUpdate {
    pub color: Option<String>,
    pub note: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BatchCreateResponse {
    pub highlights: Vec<Highlight>,
    pub created: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteResponse {
    pub success: bool,
    pub deleted: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportResponse {
    pub content: String,
    pub filename: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FileExistsResponse {
    pub exists: bool,
    pub file_path: String,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum OrderBy {
    LastOpened,
    Created,
    Title,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ExportFormat {
    Markdown,
    Json,
    Text,
}

// =============================================================================
// Error Types
// =============================================================================

#[derive(Debug, Clone, thiserror::Error)]
pub enum DomainError {
    #[error("Entity not found: {0}")]
    NotFound(String),

    #[error("Validation error: {0}")]
    Validation(String),

    #[error("Storage error: {0}")]
    Storage(String),

    #[error("TTS error: {0}")]
    Tts(String),

    #[error("File system error: {0}")]
    FileSystem(String),
}

// =============================================================================
// Port Traits
// =============================================================================

/// DocumentRepository Port
///
/// Manages document persistence operations.
/// Implemented by: SqliteDocumentRepo, MockDocumentRepo
#[cfg_attr(test, automock)]
#[async_trait]
pub trait DocumentRepository: Send + Sync {
    /// Add a new document to the library
    async fn add(
        &self,
        file_path: &str,
        title: Option<&str>,
        page_count: Option<i32>,
    ) -> Result<Document, DomainError>;

    /// Get a document by its content-hash ID
    async fn get_by_id(&self, id: &str) -> Result<Option<Document>, DomainError>;

    /// Get a document by its file path
    async fn get_by_path(&self, path: &str) -> Result<Option<Document>, DomainError>;

    /// List all documents with optional ordering and pagination
    async fn list(
        &self,
        order: OrderBy,
        limit: Option<i32>,
        offset: Option<i32>,
    ) -> Result<Vec<Document>, DomainError>;

    /// Update reading progress for a document
    async fn update_progress(
        &self,
        id: &str,
        page: i32,
        scroll: Option<f64>,
        tts_chunk: Option<&str>,
    ) -> Result<Document, DomainError>;

    /// Update document title
    async fn update_title(&self, id: &str, title: &str) -> Result<Document, DomainError>;

    /// Relocate a document to a new file path
    async fn relocate(&self, id: &str, new_path: &str) -> Result<Document, DomainError>;

    /// Remove a document from the library
    async fn remove(&self, id: &str) -> Result<(), DomainError>;

    /// Check if a file exists at the given path
    async fn check_file_exists(&self, path: &str) -> Result<FileExistsResponse, DomainError>;
}

/// HighlightRepository Port
///
/// Manages highlight persistence operations.
/// Implemented by: SqliteHighlightRepo, MockHighlightRepo
#[cfg_attr(test, automock)]
#[async_trait]
pub trait HighlightRepository: Send + Sync {
    /// Create a new highlight
    async fn create(&self, input: HighlightCreate) -> Result<Highlight, DomainError>;

    /// Create multiple highlights in a batch
    async fn batch_create(
        &self,
        inputs: Vec<HighlightCreate>,
    ) -> Result<BatchCreateResponse, DomainError>;

    /// Get a highlight by ID
    async fn get_by_id(&self, id: &str) -> Result<Option<Highlight>, DomainError>;

    /// List all highlights for a specific page
    async fn list_for_page(
        &self,
        document_id: &str,
        page_number: i32,
    ) -> Result<Vec<Highlight>, DomainError>;

    /// List all highlights for a document
    async fn list_for_document(&self, document_id: &str) -> Result<Vec<Highlight>, DomainError>;

    /// Update a highlight's color or note
    async fn update(&self, id: &str, input: HighlightUpdate) -> Result<Highlight, DomainError>;

    /// Delete a highlight
    async fn delete(&self, id: &str) -> Result<(), DomainError>;

    /// Delete all highlights for a document
    async fn delete_for_document(&self, document_id: &str) -> Result<DeleteResponse, DomainError>;

    /// Export highlights in various formats
    async fn export(
        &self,
        document_id: &str,
        format: ExportFormat,
    ) -> Result<ExportResponse, DomainError>;
}

/// TtsEngine Port
///
/// Text-to-speech engine operations.
/// Implemented by: NativeTtsAdapter, ElevenLabsAdapter, MockTtsAdapter
#[cfg_attr(test, automock)]
#[async_trait]
pub trait TtsEngine: Send + Sync {
    /// Initialize the TTS engine
    async fn init(&self) -> Result<TtsInitResponse, DomainError>;

    /// List available voices
    async fn list_voices(&self) -> Result<Vec<VoiceInfo>, DomainError>;

    /// Speak a single text chunk
    async fn speak(&self, text: &str, chunk_id: Option<&str>) -> Result<(), DomainError>;

    /// Speak multiple chunks sequentially (for long content)
    async fn speak_long(
        &self,
        chunks: Vec<(String, String)>, // (id, text)
    ) -> Result<(), DomainError>;

    /// Stop all speech
    async fn stop(&self) -> Result<(), DomainError>;

    /// Pause current speech
    async fn pause(&self) -> Result<(), DomainError>;

    /// Resume paused speech
    async fn resume(&self) -> Result<(), DomainError>;

    /// Set the active voice
    async fn set_voice(&self, voice_id: &str) -> Result<(), DomainError>;

    /// Set playback rate (0.5 - 3.0)
    async fn set_rate(&self, rate: f64) -> Result<(), DomainError>;

    /// Get current TTS state
    async fn get_state(&self) -> Result<TtsState, DomainError>;

    /// Check engine capabilities
    async fn check_capabilities(&self) -> Result<TtsCapabilities, DomainError>;
}

/// SettingsRepository Port
///
/// User settings persistence.
/// Implemented by: SqliteSettingsRepo, MockSettingsRepo
#[cfg_attr(test, automock)]
#[async_trait]
pub trait SettingsRepository: Send + Sync {
    /// Get a setting value by key
    async fn get(&self, key: &str) -> Result<Option<String>, DomainError>;

    /// Set a setting value
    async fn set(&self, key: &str, value: &str) -> Result<(), DomainError>;

    /// Get all settings
    async fn get_all(&self) -> Result<Vec<(String, String)>, DomainError>;

    /// Delete a setting
    async fn delete(&self, key: &str) -> Result<(), DomainError>;

    /// Set multiple settings at once
    async fn set_batch(&self, settings: Vec<(String, String)>) -> Result<(), DomainError>;
}

/// FileSystem Port
///
/// File system operations (restricted by Tauri security model).
/// Implemented by: TauriFileSystemAdapter, MockFileSystemAdapter
#[cfg_attr(test, automock)]
#[async_trait]
pub trait FileSystemPort: Send + Sync {
    /// Read file contents
    async fn read_file(&self, path: &str) -> Result<Vec<u8>, DomainError>;

    /// Write file contents
    async fn write_file(&self, path: &str, data: &[u8]) -> Result<(), DomainError>;

    /// Check if file exists
    async fn exists(&self, path: &str) -> Result<bool, DomainError>;

    /// Compute SHA-256 hash of file contents
    async fn hash_file(&self, path: &str) -> Result<String, DomainError>;
}

// =============================================================================
// Application Service Traits (Use Cases)
// =============================================================================

/// LibraryService
///
/// Use cases for document library management.
#[async_trait]
pub trait LibraryService: Send + Sync {
    async fn open_document(&self, file_path: &str) -> Result<Document, DomainError>;
    async fn close_document(&self, id: &str) -> Result<(), DomainError>;
    async fn get_recent_documents(&self, limit: Option<i32>) -> Result<Vec<Document>, DomainError>;
    async fn remove_document(&self, id: &str) -> Result<(), DomainError>;
}

/// ReadingService
///
/// Use cases for document reading and navigation.
#[async_trait]
pub trait ReadingService: Send + Sync {
    async fn go_to_page(&self, document_id: &str, page: i32) -> Result<(), DomainError>;
    async fn update_scroll(&self, document_id: &str, scroll: f64) -> Result<(), DomainError>;
    async fn bookmark_position(&self, document_id: &str) -> Result<(), DomainError>;
}

/// TtsService
///
/// Use cases for text-to-speech playback.
#[async_trait]
pub trait TtsService: Send + Sync {
    async fn start_reading(
        &self,
        document_id: &str,
        from_chunk: Option<&str>,
    ) -> Result<(), DomainError>;
    async fn stop_reading(&self) -> Result<(), DomainError>;
    async fn toggle_pause(&self) -> Result<(), DomainError>;
    async fn skip_to_chunk(&self, chunk_id: &str) -> Result<(), DomainError>;
}
